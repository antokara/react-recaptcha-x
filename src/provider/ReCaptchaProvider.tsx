import * as React from 'react';
import { Context } from './Context';
import { IContext } from './IContext';

declare global {
  // tslint:disable-next-line:interface-name (@see https://github.com/Microsoft/TypeScript/issues/19816)
  interface Window {
    GoogleReCaptcha_onload?: Function;
  }
}

type TProps = {
  siteKeyV2?: string;
  siteKeyV3?: string;
  langCode?: string;
  hideV3Badge?: boolean;
  children: React.ReactNode | React.ReactNodeArray;
};
type TState = {
  loaded: boolean;
};

/**
 * a Provider which must be used once per application,
 * to include the Google reCAPTCHA JS API
 */
class ReCaptchaProvider extends React.Component<TProps, TState> {
  constructor(props: TProps) {
    super(props);
    this.state = {
      loaded: false
    };
    this.onLoadHandler = this.onLoadHandler.bind(this);
  }

  public componentDidMount(): void {
    const {
      siteKeyV3 = 'explicit',
      langCode = '',
      hideV3Badge = false
    } = this.props;
    const scriptDataId: string = 'react-recaptcha-v3-v2-script';
    // avoid loading again if previously loaded...
    if (document.querySelector(`script[data-id="${scriptDataId}"]`) === null) {
      // load the Google reCAPTCHA JS API script tag.
      // We cannot dynamically import because
      // there are no CORS headers and the FETCH will fail if we try...
      window.GoogleReCaptcha_onload = this.onLoadHandler;
      const script: HTMLScriptElement = document.createElement('script');
      script.setAttribute('data-id', scriptDataId);
      script.setAttribute(
        'src',
        `https://www.google.com/recaptcha/api.js?render=${siteKeyV3}&onload=GoogleReCaptcha_onload&hl=${langCode}`
      );
      script.setAttribute('async', 'true');
      script.setAttribute('defer', 'true');
      document.body.appendChild(script);
    } else if (window.GoogleReCaptcha_onload) {
      // in case the provider was previously mounted, then unmounted before invoking
      // our Google reCAPTCHA onload handler, replace the onload handler with the provider's
      // so that we can capture then event,set the loaded state and propagate the event...
      window.GoogleReCaptcha_onload = this.onLoadHandler;
    } else {
      // in case the provider was previously mounted, then unmounted but
      // had also invoked our Google reCAPTCHA onload handler, only then, change to already loaded
      this.setState({ loaded: true });
    }

    // in the case user wants to hide the reCaptchaV3 badge and
    // if it wasn't previously appended...
    // @see https://developers.google.com/recaptcha/docs/faq
    const styleDataId: string = 'react-recaptcha-v3-v2-style';
    if (
      hideV3Badge &&
      document.querySelector(`style[data-id="${styleDataId}"]`) === null
    ) {
      const style: HTMLStyleElement = document.createElement('style');
      // tslint:disable-next-line:no-inner-html (this is safe, there is no user input)
      style.innerHTML = '.grecaptcha-badge{display: none;}';
      style.setAttribute('data-id', styleDataId);
      document.body.appendChild(style);
    }
  }

  /**
   * in case we unmount while the Google reCAPTCHA JS API was loading,
   * meaning, our GoogleReCaptcha_onload hasn't been triggered yet,
   * make sure when it does trigger, it does not throw and is handled properly...
   */
  public componentWillUnmount(): void {
    if (window.GoogleReCaptcha_onload) {
      window.GoogleReCaptcha_onload = (): void => {
        // remove the onload handler since it got used
        delete window.GoogleReCaptcha_onload;
      };
    }
  }

  public render(): React.ReactNode {
    const { children, siteKeyV3, siteKeyV2 } = this.props;
    const { loaded } = this.state;
    const contextValue: IContext = {
      siteKeyV2,
      siteKeyV3,
      loaded
    };

    return <Context.Provider value={contextValue}>{children}</Context.Provider>;
  }

  /**
   * invoked when the Google reCAPTCHA JS API is loaded
   * to set the loaded state
   * see componentDidMount
   */
  private onLoadHandler(): void {
    // remove the onload handler since it got used
    delete window.GoogleReCaptcha_onload;
    // propagate the loaded event/state
    this.setState({ loaded: true });
  }
}

export { ReCaptchaProvider };
