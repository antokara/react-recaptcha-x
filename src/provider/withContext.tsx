/* eslint-disable @typescript-eslint/ban-types */
import * as React from 'react';
import { Context } from './Context';
import { IContext } from './IContext';

/**
 * context consumer HOC type
 *
 * a function that takes in the component and
 * returns the same component but with the context passed as a prop
 */
type TWithContext = <T extends object>(
  Component: React.ComponentType<T>
) => React.ComponentType<T>;

/**
 * consumer child type
 *
 * a function that is invoked with the context value and
 * returns the consumer component
 */
type TConsumerChild = (value: IContext) => React.ReactElement;

/**
 * context consumer child generator type
 *
 * takes the component with its props and
 * returns a function that creates the consumer child component
 */
type TConsumerChildGenerator = <T extends object>(
  Component: React.ComponentType<T>,
  props: T
) => TConsumerChild;

/**
 * context consumer child generator
 *
 * - gets the component to wrap and the props of the HOC.
 * - returns a function that accepts the Context value
 * - finally, returns the Component with both
 *  - ownProps of the Component
 *  - providerContext prop
 */
const ConsumerChildGenerator: TConsumerChildGenerator =
  <T extends object>(
    Component: React.ComponentType<T>,
    props: T
  ): TConsumerChild =>
  (value: IContext): React.ReactElement =>
    <Component {...props} providerContext={value} />;

/**
 * context consumer HOC
 *
 * a function that takes in the component and
 * returns the same component but with the context passed as a prop
 */
const withContext: TWithContext = <T extends object>(
  Component: React.ComponentType<T>
): React.ComponentType<T> =>
  class WithContext extends React.Component<T> {
    public render(): React.ReactNode {
      return (
        <Context.Consumer>
          {ConsumerChildGenerator(Component, this.props)}
        </Context.Consumer>
      );
    }
  };

export { withContext };
