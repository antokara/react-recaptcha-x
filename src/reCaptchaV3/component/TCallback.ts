import { TRefreshToken } from './TRefreshToken';

/**
 * callback type
 *  - when called without arguments, it means requesting token in progress
 *  - when called with string (token) and refreshToken, it means token retrieved
 *
 * refreshToken can be deferred and invoked whenever to generate a new token
 */
export type TCallback = (
  token: string | void,
  refreshToken: TRefreshToken | void
) => void;
