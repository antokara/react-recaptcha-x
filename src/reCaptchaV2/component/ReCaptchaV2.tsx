import * as React from 'react';
import { IConsumer } from 'src/provider/IConsumer';
import { ESize } from './ESize';
import { ETheme } from './ETheme';
import { IProps } from './IProps';
import { IState } from './IState';

/**
 * a React reCAPTCHA version 2 component
 */
class ReCaptchaV2 extends React.Component<IProps & IConsumer, IState> {
  public static defaultProps: Pick<IProps, 'theme' | 'size' | 'tabindex'> = {
    theme: ETheme.Light,
    size: ESize.Normal,
    tabindex: 0
  };

  public constructor(props: IProps & IConsumer) {
    super(props);

    // validate the presence of the siteKeyV2
    const { siteKeyV2 } = this.props.providerContext;
    if (!siteKeyV2) {
      throw new Error(
        'The prop "siteKeyV2" must be set on the ReCaptchaProvider before using the ReCaptchaV2 component'
      );
    }

    this.state = {
      ref: React.createRef<HTMLDivElement>(),
      widgetId: undefined,
      token: undefined,
      expired: false,
      error: false
    };

    this.successCallback = this.successCallback.bind(this);
    this.expiredCallback = this.expiredCallback.bind(this);
    this.errorCallback = this.errorCallback.bind(this);
    this.checkComponentDidUpdate = this.checkComponentDidUpdate.bind(this);
  }

  public componentDidUpdate(prevProps: IProps & IConsumer): void {
    this.checkComponentDidUpdate(this.state.ref, this.props, prevProps);
  }

  /**
   * if js api is loaded and was not previously loaded,
   * attempt to render the widget
   */
  public checkComponentDidUpdate(
    ref: React.RefObject<HTMLDivElement>,
    props: IProps & IConsumer,
    prevProps: IProps & IConsumer
  ): void {
    const { loaded, siteKeyV2 } = props.providerContext;
    const { theme, size, tabindex } = props;
    // make sure the providerContext changed to loaded,
    // we have a ref.current defined (happens in constructor) and
    // the siteKeyV2 has been defined as well.
    if (
      prevProps.providerContext.loaded !== loaded &&
      loaded &&
      ref.current &&
      siteKeyV2
    ) {
      // render the widget and store the returned widget id in the state
      this.setState({
        expired: false,
        error: false,
        widgetId: grecaptcha.render(ref.current, {
          sitekey: siteKeyV2,
          callback: this.successCallback,
          'expired-callback': this.expiredCallback,
          'error-callback': this.errorCallback,
          theme,
          size,
          tabindex
        })
      });
    }
  }

  public render(): JSX.Element {
    const { ref } = this.state;
    const {
      theme,
      callback,
      size,
      tabindex,
      providerContext,
      ...other
    } = this.props;

    return <div ref={ref} {...other} />;
  }

  /**
   * invoked by the widget when the user submits a successful response.
   * The g-recaptcha-response token is provided.
   */
  private successCallback(token: string): void {
    const { callback } = this.props;
    this.setState(
      {
        token,
        expired: false,
        error: false
      },
      () => {
        // invoke callback with token, to signify success and pass the token
        callback(token);
      }
    );
  }

  /**
   * invoked by the widget when the response expired and the user needs to re-verify
   */
  private expiredCallback(): void {
    const { callback } = this.props;
    this.setState(
      {
        expired: true
      },
      () => {
        // invoke callback with false, to signify expiration
        callback(false);
      }
    );
  }

  /**
   * invoked by the widget when an error occurrs (usually network connectivity) and
   * it cannot continue until connectivity is restored.
   */
  private errorCallback(): void {
    const { callback } = this.props;
    this.setState(
      {
        error: true
      },
      () => {
        // invoke callback with error, to signify error
        callback(new Error());
      }
    );
  }
}

export { ReCaptchaV2, IProps };
