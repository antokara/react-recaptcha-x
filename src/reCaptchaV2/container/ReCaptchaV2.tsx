import * as React from 'react';
import { withContext } from '../../provider/withContext';
import { IProps } from '../../reCaptchaV2/component/IProps';
import { ReCaptchaV2 as Component } from '../../reCaptchaV2/component/ReCaptchaV2';

// provide access to the provider context
const ReCaptchaV2: React.ComponentType<IProps> = withContext(Component);

export { ReCaptchaV2 };
