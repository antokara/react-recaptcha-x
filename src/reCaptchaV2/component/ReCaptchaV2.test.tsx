import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { IContext } from 'src/provider/IContext';
import { ESize } from './ESize';
import { ETheme } from './ETheme';
import { ReCaptchaV2 } from './ReCaptchaV2';
import { TCallback } from './TCallback';

describe('ReCaptchaV2 component', () => {
  let renderer: TestRenderer.ReactTestRenderer;
  let callback: TCallback;
  let providerContext: IContext;

  beforeEach(() => {
    callback = jest.fn();
    providerContext = {
      siteKeyV2: 'test',
      siteKeyV3: undefined,
      loaded: false
    };

    renderer = TestRenderer.create(
      <ReCaptchaV2
        callback={callback}
        theme={ETheme.Light}
        tabindex={0}
        size={ESize.Normal}
        providerContext={providerContext}
      />
    );
  });

  it('matches the snapshot', () => {
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
