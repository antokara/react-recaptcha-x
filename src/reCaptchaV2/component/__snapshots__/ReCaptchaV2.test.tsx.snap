// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReCaptchaV2 component matches the snapshot 1`] = `<div />`;
